packages/mattermost-redux/src/utils/apps.ts

```diff
@@ -1,116 +1,189 @@
-    }
+        switch (topLocation) {
+        case AppBindingLocations.COMMAND: {
+            if (b.label.match(/ |\t/)) {
+                toRemove.unshift(i);
+                return;
+            }
+
+            if (usedLabels[b.label]) {
+                toRemove.unshift(i);
+                return;
+            }
+            break;
+        }
+        case AppBindingLocations.CHANNEL_HEADER_ICON: {
+            // First level of channel header icons must have an icon to show as the icon
+            if (!b.icon && depth === 0) {
+                toRemove.unshift(i);
+                return;
+            }
+            break;
+        }
+        }
 
-    // Trim branches without calls
-    if (!binding.call) {
-        binding.bindings = binding.bindings?.filter((v) => v.call);
-    }
+        if (b.bindings?.length) {
+            cleanBinding(b, topLocation, depth + 1);
+
+            // Remove invalid branches
+            if (!b.bindings?.length) {
+                toRemove.unshift(i);
+                return;
+            }
+        } else {
+            // Remove leaves without a call
+            if (!b.call) {
+                toRemove.unshift(i);
+                return;
+            }
+
+            // Remove leaves without app id
+            if (!b.app_id) {
+                toRemove.unshift(i);
+                return;
+            }
+        }
 
-    // Pull up app_id if needed
-    if (binding.bindings?.length && !binding.app_id) {
-        binding.app_id = binding.bindings[0].app_id;
-    }
+        usedLabels[b.label] = true;
+    });
 
-    // Pull up call if needed
-    if (binding.bindings?.length && !binding.call) {
-        binding.call = binding.bindings[0].call;
-    }
+    toRemove.forEach((i) => {
+        binding.bindings?.splice(i, 1);
+    });
 }
 
-export function validateBindings(binding?: AppBinding) {
-    filterInvalidChannelHeaderBindings(binding);
-    filterInvalidCommands(binding);
-    filterInvalidPostMenuBindings(binding);
-    binding?.bindings?.forEach(fillAndTrimBindingsInformation);
+export function validateBindings(bindings: AppBinding[] = []): AppBinding[] {
+    const channelHeaderBindings = bindings?.filter((v) => v.location === AppBindingLocations.CHANNEL_HEADER_ICON);
+    const postMenuBindings = bindings?.filter((v) => v.location === AppBindingLocations.POST_MENU_ITEM);
+    const commandBindings = bindings?.filter((v) => v.location === AppBindingLocations.COMMAND);
+
+    channelHeaderBindings.forEach((v) => cleanBinding(v, AppBindingLocations.CHANNEL_HEADER_ICON, 0));
+    postMenuBindings.forEach((v) => cleanBinding(v, AppBindingLocations.POST_MENU_ITEM, 0));
+    commandBindings.forEach((v) => cleanBinding(v, AppBindingLocations.COMMAND, 0));
+
+    const hasBindings = (v: AppBinding) => v.bindings?.length;
+    return postMenuBindings.filter(hasBindings).concat(channelHeaderBindings.filter(hasBindings), commandBindings.filter(hasBindings));
```

Comment:
Nit, but I would prefer `b` instead of `v` to represent a `binding`

Reply:
I was going with `v` for `value` but you are right, `b` for `binding` looks better here.